lib/msal-core/src/Utils.ts

```diff
@@ -567,48 +568,220 @@ export class Utils {
+      ssoType = SSOTypes.SID;
+    }
+    // loginHint from request
+    else if (request.loginHint) {
+      ssoData = request.loginHint;
+      ssoType = SSOTypes.LOGIN_HINT;
+    }
+    // adalIdToken retrieved from cache
+    else if (idTokenObject) {
       if (idTokenObject.hasOwnProperty(Constants.upn)) {
-        extraQueryParameters = this.urlRemoveQueryStringParameter(extraQueryParameters, Constants.login_hint);
-        extraQueryParameters = this.urlRemoveQueryStringParameter(extraQueryParameters, Constants.domain_hint);
-        if (extraQueryParameters) {
-          return extraQueryParameters += "&" + Constants.login_hint + "=" + idTokenObject.upn + "&" + Constants.domain_hint + "=" + Constants.organizations;
-        }
-        else {
-          return extraQueryParameters = "&" + Constants.login_hint + "=" + idTokenObject.upn + "&" + Constants.domain_hint + "=" + Constants.organizations;
-        }
+        ssoData = idTokenObject.upn;
+        ssoType = SSOTypes.ID_TOKEN;
       }
       else {
-        extraQueryParameters = this.urlRemoveQueryStringParameter(extraQueryParameters, Constants.domain_hint);
-        if (extraQueryParameters) {
-          return extraQueryParameters += "&" + Constants.domain_hint + "=" + Constants.organizations;
+        ssoData = null;
+        ssoType = SSOTypes.ORGANIZATIONS;
+      }
+    }
+
+    extraQueryParameters = this.processExtraQueryParameters(extraQueryParameters, ssoData, ssoType);
+    return extraQueryParameters;
+  }
+
+  /**
+   * Remove the SSO data from extraQueryParams if passed and
+   * construct the string "extraQueryParameters" needed to be sent to the server
+   * @param extraQueryParameters
+   * @param ssoData
+   * @param ssoType
+   */
+  static processExtraQueryParameters(extraQueryParameters: string, ssoData: string, ssoType: string): string {
+    extraQueryParameters = this.removeSSOParams(extraQueryParameters, ssoType);
+
+    if (extraQueryParameters) {
+        extraQueryParameters += this.generateSSOUrlParameter(ssoData, ssoType);
+    }
+    else {
+        extraQueryParameters = this.generateSSOUrlParameter(ssoData, ssoType);
+    }
+
+    return extraQueryParameters;
+  }
+
+  /**
+   * remove the exisiting SSOData from the extraQueryParameters
+   * @param sid
+   */
+  static removeSSOParams(extraQueryParameters: string, type: string): string {
+
+    let ssoType: string;
+    if (type === SSOTypes.ID_TOKEN) {
+        ssoType = SSOTypes.LOGIN_HINT;
+    }
+    else {
+        ssoType = type;
+    }
+
+    extraQueryParameters = this.urlRemoveQueryStringParameter(extraQueryParameters, ssoType);
+    if (type === SSOTypes.ID_TOKEN) {
+        extraQueryParameters = this.urlRemoveQueryStringParameter(extraQueryParameters, SSOTypes.DOMAIN_HINT);
+    }
+
+    return extraQueryParameters;
+  }
+
+  /**
+   * Add SID to extraQueryParameters
+   * @param sid
+   */
+  static generateSSOUrlParameter(ssoData: string, ssoType: string): string {
```

Comment:
it's little odd to have the value as the first argument, and name the 2nd argument.

Reply:
Changes made.